### ISSUE TITLE
Hybrid Deep Learning Framework for Carnatic Raga Recognition using Audio Feature Fusion and Explainable AI

### ISSUE THREAD

**Issue by varnika211005** [Niketkumardheeryan/ML-CaPsule#1678]
## Project Overview

This project aims to develop an advanced AI-based Carnatic Raga Recognition System using hybrid Machine Learning and Deep Learning techniques. Unlike traditional classification systems that rely only on static spectral features, this framework will combine temporal audio dynamics, feature fusion, and explainable AI for improved raga identification.

The system will analyze musical characteristics such as MFCCs, Chroma features, Spectral Centroid, Spectral Bandwidth, and temporal frequency variations to classify Carnatic ragas with higher robustness and interpretability.

---

## Proposed Enhancements

* Deep learning-based audio sequence modeling
* Temporal feature learning using BiLSTM/CNN architectures
* Feature fusion of spectral and rhythmic descriptors
* Explainable AI visualizations for prediction interpretation
* Comparative analysis between traditional ML and DL models
* Confusion matrix and feature importance visualization
* Model export for real-time deployment

---

### Machine Learning

* Random Forest
* SVM
* XGBoost

### Deep Learning

* CNN
* BiLSTM
* Hybrid CNN-BiLSTM architecture

---

## Advanced Features

* Audio augmentation for better generalization
* Class imbalance handling
* Hyperparameter tuning
* Real-time audio prediction support
* Explainable AI using SHAP/LIME

---

## Future Enhancements

* Real-time/live audio input support for instant raga prediction
* Microphone-based raga recognition system
* Deep learning-based end-to-end audio classification
* Mobile/web deployment for live music analysis
* Transformer-based audio sequence modeling
* Multi-label emotion and raga mood detection
* Real-time concert performance analysis
* Integration with music learning applications
* Support for additional Carnatic and Hindustani ragas
* Personalized AI-assisted raga training and feedback system

---

## Dataset

Carnatic raga audio feature dataset containing acoustic descriptors such as:

* MFCCs
* Chroma STFT
* RMSE
* Spectral Centroid
* Spectral Bandwidth

---

## Expected Outcome

The project aims to build an intelligent music classification framework capable of accurately identifying Carnatic ragas while also providing interpretability into the model’s predictions.

---

## Tech Stack

* Python
* TensorFlow / PyTorch
* Scikit-learn
* Librosa
* NumPy
* Pandas
* Matplotlib
* Seaborn
* SHAP / LIME

I would love to contribute this project under GSSoC 2026 in the ML-Capsule repository.

**Comment by varnika211005** [comment]
@Niketkumardheeryan  Hi, I would love to contribute to this project under GSSoC 2026. I’m very interested in AI/ML and would be grateful if you could review this proposal and consider assigning it to me.

Thank you!

**Comment by YashYS04** [comment]
Hi @varnika211005 , I would like to do this task, assign it to me.

**Comment by varnika211005** [comment]
# Project Progress Update

## Current Progress / Implementation Update

I have already started working on the implementation of the Carnatic Raga Recognition System and completed several core components of the project.

### Completed Work

* Dataset preprocessing and balancing
* Feature extraction pipeline setup
* Feature scaling and normalization
* Label encoding for raga classification
* Training and evaluation of multiple ML models:

  * Random Forest
  * Support Vector Machine (SVM)
  * K-Nearest Neighbors (KNN)
* Comparative model accuracy analysis
* Confusion matrix visualization
* Feature importance analysis
* Model serialization using Joblib

### Current Results

* Successfully trained classification models on Carnatic raga audio feature datasets
* Achieved strong classification performance using Random Forest
* Generated visual evaluation metrics including:

  * Confusion Matrix
  * Model Comparison Graphs
  * Feature Importance Charts

### Planned Extensions

* Hybrid CNN-BiLSTM deep learning architecture
* Explainable AI integration using SHAP/LIME
* Real-time/live audio prediction support
* Microphone-based raga recognition
* Streamlit/Web deployment
* Temporal audio sequence learning
* Audio augmentation and advanced preprocessing

### Additional Goal

I also plan to showcase the implementation through an interactive demo interface capable of identifying ragas from uploaded or live audio inputs.

I would love to further develop and contribute this project under GSSoC 2026 in the ML-CaPsule repository.

## Performance Evaluation and Visualization

### 1. Confusion Matrix — Random Forest Classifier
<img width="698" height="541" alt="Image" src="https://github.com/user-attachments/assets/39d182e3-62b6-469e-b1a2-fd2be5e9b36b" />

### 2. Comparative Accuracy Analysis of ML Algorithms
<img width="790" height="490" alt="Image" src="https://github.com/user-attachments/assets/165fd643-c82f-4027-91d1-f9aaf20e2f1f" />

### 3. Top Audio Features Influencing Raga Classification
<img width="989" height="590" alt="Image" src="https://github.com/user-attachments/assets/6f7cb82e-578b-44ed-a115-6b36903b0128" />

**Comment by varnika211005** [comment]
@Niketkumardheeryan Hi, I’ve added the current implementation progress and evaluation results for the proposed project. I’d be grateful if you could review the proposal when possible.

I’d really love to contribute to this project under GSSoC 2026. Thank you!

**Comment by varnika211005** [comment]
## Project repository for the current implementation:

https://github.com/varnika211005/Carnatic-Raga-Recognition

I’ve uploaded the current implementation, evaluation results, and visualizations for the proposed Carnatic Raga Recognition System.

**Comment by vishisht-gupta** [comment]
i am also interested as i have strong interest in AI/ML

**Comment by varnika211005** [comment]
Hello sir,@Niketkumardheeryan

I wanted to provide a quick update regarding this issue.

I have already submitted PR #1796 containing the current implementation, including the machine learning pipeline, model evaluation, visualizations, and project documentation.

The project proposal also includes several future enhancements such as CNN-BiLSTM integration, Explainable AI (SHAP/LIME), real-time audio prediction, and deployment features. Due to my ongoing semester-end examinations, I have not yet been able to complete these advanced extensions.

My exams will conclude shortly, after which I intend to actively continue working on the project and implement the proposed enhancements.

I wanted to share this update since the issue is approaching the auto-unassign period. Please let me know if there are any suggestions or modifications required from my side.

Thank you for your time and guidance.

**Comment by vishisht-gupta** [comment]
Hey i can help in your problem  i have the knowledge how to train and
connect them . If you want i can share the the some good project to you
